src/coreclr/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -7615,6 +7619,64 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
     return S_OK;
 }
 
+HRESULT ProfToEEInterfaceImpl::EnumerateGCHeapObjects(ObjectCallback callback, void* callbackState)
+{
+    CONTRACTL
+    {
+        // Yay!
+        NOTHROW;
+
+        // Suspending EE is contracted to trigger GC through ThreadSuspend::SuspendAllThreads
+        GC_TRIGGERS;
+
+        // Yay!
+        MODE_ANY;
+
+        // Suspending EE will acquire the ThreadStore Lock via ThreadSuspend::LockThreadStore.
+        CAN_TAKE_LOCK;
+
+        // Yay!
+        EE_THREAD_NOT_REQUIRED;
+    }
+    CONTRACTL_END;
+
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(kP2EEAllowableAfterAttach | kP2EETriggers,
+        (LF_CORPROF,
+        LL_INFO1000,
+        "**PROF: EnumerateGCHeapObjects.\n"));
+
+    if (callback == nullptr)
+    {
+        return E_INVALIDARG;
+    }
+
+    if (!g_fEEStarted)
+    {
+        return CORPROF_E_RUNTIME_UNINITIALIZED;
+    }
+
+    if (!ThreadSuspend::SysIsSuspendInProgress() && (ThreadSuspend::GetSuspensionThread() == 0))
+    {
+        ThreadSuspend::SuspendEE(ThreadSuspend::SUSPEND_REASON::SUSPEND_FOR_PROFILER);
+    }
+
+    // Suspending EE ensures safe object inspection. We permit the GC Heap walk callback to
+    // invoke ICorProfilerInfo APIs guarded by AllowObjectInspection by toggling fGCInProgress.
+    g_profControlBlock.fGCInProgress = TRUE;
+
+    // Leveraging a direct callback instead of a ICorProfilerCallback API avoids the performance overhead of
```

Comment:
I do not see what we are trying to protect against here. The profiler should prevent itself from getting unloaded when it calls EnumerateGCHeapObjects. It is not the runtime job to prevent the profiler from getting unloaded in this situation.

Reply:
We don't need to guard against unloading, we need to set the profiler thread flags so the profiler can call the `PROFILER_TO_CLR_ENTRYPOINT_SYNC` methods and do object inspection in the callback

Reply:
Yes, your description makes sense.

This calls `DoProfilerCallback`. `DoProfilerCallback` is mostly concerned with calling all attached profilers that is not something that we want to do here.

Is there a better - more direct and lightweight - way to set the profiler flags?

Reply:
There might be a way through `SetProfilerCallbackStateFlags`, since https://github.com/dotnet/runtime/blob/6b4b371c44ea14536f860b75b5a1553605f49c82/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L244 is really what is blocking the non-EEToProfInterfaceImpl callback from invoking synchronous APIs guarded with the logic under the`PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX` macro. But didn't go down that route, and opted to use a "shell" EEToProfInterfaceImpl wrapper helper.

Reply:
The wrapper is fine. The part that I do not understand is what you get by going through `DoProfilerCallback`.

What is not going to work if you call EnumerateGCHeapObjectsCallback directly from here instead of going through `DoProfilerCallback`?

Reply:
I originally mimicked the pattern to invoke EEToProfInterfaceImpl method, not realizing at the time that DoProfilerCallback would iterate over all attached profilers.

Right now, given that the macro for `PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX` uses members of the EEToProfIntefaceImpl class + `this`, I wouldn't be able to invoke it statically. I didn't see (from a glance) any instances of calling EEToProfInterfaceImpl methods from ProfToEEInterfaceImpl APIs that weren't calling through `g_profControlBlock`, so I naively followed suit there at the time.

What would be the best way to grab the "current" profiler? I'm wondering if something like ` mainProfilerInfo.pProfInterface.Load()` makes sense, or if there would ever be an instance where a non-main profiler would invoke `EnumerateGCHeapObjects`.

Reply:
> What would be the best way to grab the "current" profiler?

Can you get it from `ProfToEEInterfaceImpl::m_pProfilerInfo`?

Reply:
Used m_pProfilerInfo to grab the EEToProfInterfaceImpl and invoked the callback `EnumerateGCHeapObjectsCallback` directly